src/Raven.Client/Documents/Session/SessionInfo.cs

```diff
@@ -10,53 +10,62 @@ public class SessionInfo
         [ThreadStatic]
         private static int _clientSessionIdCounter;
 
-        private int _clientSessionId;
+        private int? _sessionId;
         private bool _sessionIdUsed;
         private readonly int _loadBalancerContextSeed;
+        private readonly bool _loadBalancerContextSelectorSet;
+        private readonly InMemoryDocumentSessionOperations _session;
 
         public int SessionId
         {
             get
             {
+                if (_sessionId == null)
+                    SetContext(_session.Conventions.LoadBalancerPerSessionContextSelector?.Invoke(_session.DatabaseName));
+
                 _sessionIdUsed = true;
-                return _clientSessionId;
+
+                return _sessionId.Value;
             }
         }
 
+        internal bool HasSessionId
+        {
+            get => _sessionId.HasValue || _loadBalancerContextSelectorSet;
```

Comment:
The second time this is called, is will be true, because we set `_sessionId` from `_clientSessionIdCounter`

Reply:
Done.